che-plugin-openshift-ext-client/src/main/java/org/eclipse/che/ide/ext/openshift/client/deploy/_new/NewApplicationPresenter.java

```diff
@@ -418,31 +429,101 @@ public void onImageStreamChanged(String stream) {
+                   &&view.getOpenShiftSelectedProject() != null
+                   && !osApplications.isEmpty()) {
+            for(Pair<String, String> pair : osApplications) {
+                if (pair.getFirst().equals(view.getOpenShiftSelectedProject().getMetadata().getName())
+                    && pair.getSecond().equals(osAppName)) {
+                    view.showApplicationNameError(locale.existingApplicationNameError(), null);
+                    return false;
+                }
+            }
+        }
+        view.hideApplicationNameError();
+        return true;
     }
 
-    private void validate() {
-        boolean valid = true;
+    private boolean isProjectNameValid() {
         String osProjectName = view.getOpenShiftProjectName();
-        //TODO add check for application name
-
         if (view.getMode() == CREATE_NEW_PROJECT) {
-            //TODO check existing OpenShift projects
-            if (Strings.isNullOrEmpty(osProjectName) || !osProjectName.matches("[a-z0-9]([-a-z0-9]*[a-z0-9])?")) {
-                valid = false;
+            if (!OpenShiftValidator.isProjectNameValid(osProjectName)) {
+                view.showProjectNameError(locale.invalidProjectNameError(), locale.invalidProjectNameDetailError());
+                return false;
+            }
+            if (!osProjects.isEmpty()) {
+                for(Project project : osProjects) {
+                    if (project.getMetadata().getName().equals(osProjectName)) {
+                        view.showProjectNameError(locale.existingProjectNameError(), null);
+                        return false;
+                    }
+                }
             }
         } else if (view.getMode() == SELECT_EXISTING_PROJECT) {
             if (view.getOpenShiftSelectedProject() == null) {
-                valid = false;
+                return false;
             }
         }
+        view.hideProjectNameError();
+        return true;
+    }
 
-        view.setDeployButtonEnabled(valid && view.getActiveImage() != null && view.getApplicationName() != null);
+    private boolean isVariablesListValid() {
+        List<KeyValue> variables = view.getEnvironmentVariables();
+        if (variables.isEmpty()) {
+            view.hideVariablesError();
+            return true;
+        }
+        for (KeyValue keyValue : variables) {
+            if (!(OpenShiftValidator.isEnvironmentVariableNameValid(keyValue.getKey()))) {
+                view.showVariablesError(locale.invalidVariablesError());
+                return false;
+            }
+        }
+        view.hideVariablesError();
+        return true;
+    }
+
+    private boolean isLabelListValid() {
+        List<KeyValue> labels = view.getLabels();
+        if (labels.isEmpty()) {
+            view.hideLabelsError();
+            return true;
+        }
+        for (KeyValue keyValue : labels) {
+            if (!(OpenShiftValidator.isLabelValid(keyValue.getKey())
```

Comment:
It is hard to understand this condition. I prefer to not use double negation if it is possible
```
!(OpenShiftValidator.isLabelValid(keyValue.getKey()) && OpenShiftValidator.isLabelValid(keyValue.getValue()))
```
vs
```
!OpenShiftValidator.isLabelValid(keyValue.getKey() || !OpenShiftValidator.isLabelValid(keyValue.getValue())
```